xdsl/irdl/irdl.py

```diff
@@ -712,28 +718,47 @@ class AttributeDef:
     constr: AttrConstraint
     """The attribute constraint."""
 
-    attr_name: str | None = None
-    """The attribute name, in case it is different from the field name."""
-
     def __init__(
         self,
         attr: Attribute | type[Attribute] | AttrConstraint,
-        attr_name: str | None = None,
     ):
         self.constr = attr_constr_coercion(attr)
-        self.attr_name = attr_name
 
 
 @dataclass(init=False)
-class OptAttributeDef(AttributeDef):
+class OptAttributeDef(AttributeDef, OptionalDef):
     """An IRDL attribute definition for an optional attribute."""
 
     def __init__(
         self,
         attr: Attribute | type[Attribute] | AttrConstraint,
-        attr_name: str | None = None,
     ):
-        super().__init__(attr, attr_name=attr_name)
+        super().__init__(attr)
+
+
+@dataclass(init=False)
+class PropertyDef:
```

Comment:
Properties and Attributes have very similar taste, IMO. Could we not have a class hierarchy that somehow represents that? It currently feels like they are handled quite differently.

Reply:
I changed it a bit, tell me if that works for you!